### PR TITLE
sync/rate: implements golang-like rate limiter

### PR DESCRIPTION
**Pull request by ochaton**
## Description
- sync/rate implements TokenBucket rate limiter
- client may create rate limiter with passing `limit` and `burst`
- rate limiter can have at most `burst` tokens
- rate limiter is refilled with token at rate `limit` tokens per second (can be fractional)
- ratelimit exposes 3 public methods:

	- `rate:wait([timeout=now, [n=1]])` awaits for `n` tokens within `timeout` seconds (can be fractional)
	- `rate:reserve([timestamp=now(), [n=1]])` creates reservation for `n` tokens which will be consumed at `timestamp` (non expired reservations can be cancelled)
	- `rate:allow([timestamp=now(), [n=1]])` auto reserves `n` token to be consumed at time `timestamp` (don't think someone need this method at all)

rate limiter implements TokenBucket with infinite `limit` (must set math.huge for limit). That means, that consumer is always allowed to fetch `n` tokens from rate limiter.

rate limiter implements fixed Token Bucket (without refilling) when `limit=0`. Then consumers are allowed to fetch at most `burst` tokens from rate limiter, after that rate limiter is drained and no one allowed to fetch tokens from it.

## Next Steps

- [x] Unit Tests
- [x] Update readme with some examples

### CODE REVIEW

File: sync/rate.lua [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+end
+
+
+---Creates new limit
+---@param name string? name of the limit
+---@param lim number float limit per second
+---@param burst integer? allowed burst (default=0)
+---@return sync.limit
+function limit.new(name, lim, burst)
+	if name == limit then error("Usage: limit.new([name]) or limit([name]) (not limit:new())", 2) end
```
Comment (ochaton) [inline sync/rate.lua]: fix indentations

File: sync/rate.lua [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+---@return sync.limit.reservation|false, string? error_message
+function limit:reserve(timestamp, n)
+    timestamp = tonumber(timestamp) or fiber.time()
+    n = tonumber(n) or 1
+
+    return self:_reserve(timestamp, n, math.huge)
+end
+
+
+return limit
```
Comment (ochaton) [inline sync/rate.lua]: add newline

File: sync/rate.lua [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    if restore <= 0 then return end
+
+    local tokens
+    timestamp, tokens = lim:_advance(timestamp)
+
+    tokens = math.min(lim.burst, tokens + restore)
+    lim.last_ts = timestamp
+    lim.tokens = tokens
+
+    --- think about this:
```
Comment (ochaton) [inline sync/rate.lua:54]: provide testcase

File: sync/rate.lua [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+---@field name string name of the limit
+---@field limit number limit is events per second
+---@field burst number burst is maximum number of tokens in limiter
+---@field tokens number current tokens in the limiter
+---@field last_ts number last timestamp when limiter was updated with tokens
+---@field last_event number last timestamp of rate-limited event
+local limit = {}
+
+limit.__index = limit
+limit.__tostring = function (self) return "limit<".. (self.name or 'anon') ..">" end
```
Comment (ochaton) [inline sync/rate.lua]: module name? limit -> rate

File: sync/rate.lua [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+end
+
+
+---Creates new limit
+---@param name string? name of the limit
+---@param rps number float limit per second
+---@param burst integer? allowed burst (default=0)
+---@return sync.rate
+function rate.new(name, rps, burst)
+    if name == rate then error("Usage: rate.new([name]) or limit([name]) (not limit:new())", 2) end
```
Comment (Mons) [inline sync/rate.lua]: possible should be

```
error("Usage: rate.new(name, [rps, burst]) or rate(name, [rps, burst]) (not rate:new(...))", 2)
```

File: sync/rate.lua [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    rps = tonumber(rps) or 0
+    burst = math.floor(tonumber(burst) or 0)
+
+    if rps < 0 then error("Usage: rate.new([name], rps, burst) rps must be non negative", 2) end
```
Comment (Mons) [inline sync/rate.lua]: Consider the same usage message here. name is not optional

File: sync/rate.lua [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    else
+        delta = self.rps * elapsed
+    end
+
+    return timestamp, math.min(self.burst, self.tokens + delta)
+end
+
+---Returns duration in fractinal seconds from token
+---
+---Can return `math.huge` is limit is non-positive
```
Comment (Mons) [inline sync/rate.lua]: ---Can return `math.huge` _if_ limit is non-positive

File: sync/rate.lua [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+function rate:wait(timeout, n)
+    if getmetatable(self) ~= rate then
+        error("Usage: rate:wait() (not rate.wait())", 2)
+    end
+
+    timeout = tonumber(timeout) or math.huge
+    n = tonumber(n) or 1
+
+    if n > self.burst and self.rps ~= math.huge then
+        return false, ("limit:wait(timeout=%s, n=%s) exceeds limiters burst=%s"):format(timeout, n, self.burst)
```
Comment (Mons) [inline sync/rate.lua]: limit or rate?

File: sync/rate.lua [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    end
+
+    local now = fiber.time()
+    local waitLim = math.min(timeout, math.huge)
+
+    local ok, ret = self:_reserve(now, n, waitLim)
+    if not ok then
+        local err = ret
+        return false, ("limit:wait(timeout=%s, n=%s) %s"):format(timeout, n, err)
```
Comment (Mons) [inline sync/rate.lua]: limit or rate?

**Review by Mons — CHANGES_REQUESTED** [review]

File: sync/rate.lua (hunk 5 above)
Comment (ochaton) [reply sync/rate.lua]: fixed

File: sync/rate.lua (hunk 6 above)
Comment (ochaton) [reply sync/rate.lua]: fixed

File: sync/rate.lua (hunk 8 above)
Comment (ochaton) [reply sync/rate.lua]: fixed

File: sync/rate.lua (hunk 9 above)
Comment (ochaton) [reply sync/rate.lua]: fixed, now `rate`

**Pull request merged**